src/Reflection/ReflectionFunctionAbstract.php

```diff
@@ -619,4 +619,17 @@ public function getReturnStatementsAst()
 
         return $visitor->getReturnNodes();
     }
+
+    private function findTypeFromAst($namespace, $locatedSource, $type)
+    {
+        $objectType = (new FindTypeFromAst())->__invoke(
+            $this->reflector->getContextFactory()->createForNamespace(
```

Comment:
>> Indeed, new CachedLocator($context, new AggregateSourceLocator([...]))

`$context` => `$contextFactory` ? 

--

The `ContextFactory` is invoked from the `ReflectionClass` and friends (for each method, parameters, property, etc), so I am not sure how wrapping the `SourceLocator` helps here.

If the locator cannot find the source (in the case of the `Aggregate` source locator) could we not throw a `SourceNotFound` exception?

The reasoning would be that the reflection instances would only be instantiated in one place (the reflectors) - reducing duplication and the responsibility of the locators.

If that makes sense, it is probably quite a large change so I would be happy to inject it (the `ContextFactory`) into the 5 locators that require it or now.